cf-agent/package_module.c

```diff
@@ -0,0 +1,1611 @@
+    char *request = StringFormat("%sName=%s\n%s%s",
+            options_str, name, ver ? ver : "", arch ? arch : "");
+    
+    PromiseResult res = PROMISE_RESULT_CHANGE;
+    
+    Rlist *error_message = NULL;
+    if (ReadWriteDataToPackageScript("remove", request, 
+            &error_message, wrapper) != 0)
+    {
+        Log(LOG_LEVEL_INFO, "Some error occurred while communicating with "
+            "package module while removing package.");
+        res = PROMISE_RESULT_FAIL;
+    }
+    if (error_message)
+    {
+        ParseAndLogErrorMessage(error_message);
+        res = PROMISE_RESULT_FAIL;
+        RlistDestroy(error_message);
+    }
+    
+    free(request);
+    free(options_str);
+    free(ver);
+    free(arch);
+    
+    /* We assume that at this point package is removed correctly. */
+    return res;    
+}
+
+
+static PromiseResult InstallPackageGeneric(Rlist *options, 
+        PackageType type, const char *packages_list_formatted,
+        const PackageModuleWrapper *wrapper)
+{
+    char *options_str = ParseOptions(options);
+    char *request = StringFormat("%s%s", options_str, packages_list_formatted);
+    
+    PromiseResult res = PROMISE_RESULT_CHANGE;
+    
+    const char *package_install_command = NULL;
+    if (type == PACKAGE_TYPE_FILE)
+    {
+        package_install_command = "file-install";
+    }
+    else if (type == PACKAGE_TYPE_REPO)
+    {
+        package_install_command = "repo-install";
+    }
+    else
+    {
+        /* If we end up here something bad has happened. */
+        ProgrammingError("Unsupported package type");
+    }
+    
+    Rlist *error_message = NULL;
+    if (ReadWriteDataToPackageScript(package_install_command, request, 
+                                     &error_message, wrapper) != 0)
+    {
+        Log(LOG_LEVEL_INFO, "Some error occurred while communicating with "
+            "package module while installing package.");
+        res = PROMISE_RESULT_FAIL;
+    }
+    if (error_message)
+    {
+        ParseAndLogErrorMessage(error_message);
+        res = PROMISE_RESULT_FAIL;
+        RlistDestroy(error_message);
+    }
+    
+    free(request);
+    free(options_str);
+    
+    return res;
+}
+
+PromiseResult InstallPackage(Rlist *options, 
+        PackageType type, const char *package_to_install,
+        const char *version, const char *architecture,
+        const PackageModuleWrapper *wrapper)
+{
+    Log(LOG_LEVEL_DEBUG, "Installing package '%s'", package_to_install);
+             
+    char *ver = version ? 
+        StringFormat("Version=%s\n", version) : NULL;
+    char *arch = architecture ? 
+        StringFormat("Architecture=%s\n", architecture) : NULL;
+    char *request = NULL;
+    
+    PromiseResult res = PROMISE_RESULT_CHANGE;
+    
+    if (type == PACKAGE_TYPE_FILE)
+    {
+        request = StringFormat("File=%s\n%s%s",
+            package_to_install, ver ? ver : "", arch ? arch : "");
```

Comment:
When using complicated ?: expressions for arguments, it's kinder to the reader to put each on a separate line !